File: 859. Buddy Strings/Solution.java
class Solution {
    public boolean buddyStrings(String s, String goal) {
        if (s.length() != goal.length()) return false;
        Set<Character> ch = new HashSet<>();
        char first = ' ', firstC = ' ', second = ' ', secondC = ' ';
        int l = s.length();
        for (int i = 0; i < l; i++) {
            ch.add(s.charAt(i));
            if (s.charAt(i) == goal.charAt(i)) continue;
            if (first == ' ') {
                first = s.charAt(i);
                firstC = goal.charAt(i);
            } else if (second == ' ') {
                second = s.charAt(i);
                secondC = goal.charAt(i);
            } else {
                return false;
            }
        }
        if (first == ' ') return ch.size() < l;
        if (second == ' ') return false;
        return first == secondC && firstC == second;
    }
}

File: 859. Buddy Strings/Solution.py
class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        if len(s) != len(goal): return False
        first, firstC, second, secondC = ' ', ' ', ' ', ' '
        l = len(s)
        for x, y in zip(s, goal):
            if x == y: continue
            if first == ' ':
                first, firstC = x, y
            elif second == ' ':
                second, secondC = x, y
            else:
                return False
        if first == ' ': return len(set(s)) < len(s)
        if second == ' ': return False
        return first == secondC and firstC == second
